### ISSUE TITLE
Deserializing maps and sets

### ISSUE THREAD

**Issue by sjakobi** [haskell/containers#405]
`binary`, `cereal` and `store` all deserialize `Set`s, `IntSet`s, `Map`s and `IntMap`s via `fromDistinctAscList <$> deserializeList` or even `fromList <$> deserializeList` and I've been wondering if that's the best way to do it.

The problem seems to be that this will first completely build the intermediary list before the map or set is constructed from the list.

Instead, it seems to me that the set or map should be built by repeatedly inserting a single element.

Do you agree that this should save a lot of memory allocations?

What I think is missing from `containers` to support the alternative deserialization implementations is something like a `insertDistinctMax` function that can efficiently insert an element that is known to be strictly greater than all elements in the set/map.

**Comment by treeowl** [comment]
Nope; I don't agree. `fromDistinctAscList` (which is probably what they
should all be using) is quite clever. It can reasonably be expected to be
faster and to allocate less than building the set/map one element at a
time, even taking the intermediate list into account. It might be possible
to write a version of `fromDistinctAscList` that would accept one element
at a time; that may be friendlier to the garbage collector. It would
require restructuring it to use an explicit stack, I believe. Should
definitely be doable.

**Comment by treeowl** [comment]
Actually, it would be even better to use a version that takes the
(recorded) size as an argument.

**Comment by sjakobi** [comment]
> Actually, it would be even better to use a version that takes the
> (recorded) size as an argument.

Right, I had been thinking about that too, but thought it wasn't worthwhile. Now that you mention it, it seems to me that knowing the size of the container beforehand should make it possible to build the tree without any rebalancing.

I guess what I'm looking for is something like `(Monad m, Ord a) => Int -> m a -> m (Set a)`

**Comment by sjakobi** [comment]
I can definitely recommend `fromDistinctAscList` for now: https://github.com/fpco/store/pull/96

**Comment by m-renaud** [comment]
At the minimum we should add this recommendation to the docs somewhere.

**Comment by sjakobi** [comment]
> Actually, it would be even better to use a version that takes the
> (recorded) size as an argument.

Here's a first implementation of such a function:

```

fromDistinctAscListOfSize :: Size -> [a] -> Set a
fromDistinctAscListOfSize n xs =
    case List.splitAt n' xs of
      (_, [] ) -> Tip
      (l, x:r) -> Bin n x (fromDistinctAscListOfSize n'           l)
                          (fromDistinctAscListOfSize (n - n' - 1) r)
  where n' = n `shiftR` 1
```

I haven't benchmarked it yet.

**Comment by treeowl** [comment]
The first split allocates `n/2` conses. The next splits allocate `n/4+n/4=n/2`. This smells linearithmic, which is worse than the *linear* `fromDistinctAscList`. We can do better!

**Comment by sjakobi** [comment]
A new version that avoids the conses:

```
fromDistinctAscListOfSize :: Size -> [a] -> Set a
fromDistinctAscListOfSize sz xs =
    case go sz xs of
      (s, []) -> s
      _       -> error "Set.fromDistinctAscListOfSize: List is longer than expected"
  where
    go :: Size -> [a] -> (Set a, [a])
    go n xs0
      | n <= 0    = (Tip, xs0)
      | otherwise = case go n_half xs0 of
                      (l, x:xs1) ->
                        let (r, xs2) = go (n - n_half - 1) xs1
                        in (Bin n x l r, xs2)
                      _ -> error "Set.fromDistinctAscListOfSize: List is shorter than expected"
      where n_half = n `unsafeShiftR` 1
```

No benchmarks yet.

**Comment by treeowl** [comment]
Yes, better. We should probably have something like this. But let's also go back to the beginning. What you've written here is shaped an awful lot like a parser, isn't it? The list of remaining elements is the parser state. To get back to your original idea, we need to weave in the deserialization monad:

```haskell
deserializeSet
  :: Monad m
  => Size
  -> m (Maybe a)
  -> m (Maybe (Set a))
deserializeSet size elem = ...
```

We don't depend on `transformers`, but we have our own copy of `StateT` and we can add one of `MaybeT` if that would be convenient.

**Comment by treeowl** [comment]
Roughly speaking, at least,

```haskell
deserializeSet
  :: Monad m
  => Size
  -> m (Maybe a)
  -> m (Maybe (Set a))
deserializeSet sz = runMaybeT . go sz . MaybeT
  where
    go :: Size -> MaybeT m a -> MaybeT m (Set a)
    go n elem
      | n <= 0    = pure Tip
      | otherwise = do
           l <- go n_half elem
           x <- elem
           r <- go (n - n_half - 1) elem
           pure $! Bin n x l r
        where n_half = n `unsafeShiftR` 1
```

But we should probably pull out `elem` and inline everything.

**Comment by treeowl** [comment]
Oh, what am I thinking? Much better:

```haskell
deserializeSet
  :: Monad m
  => Size
  -> m a
  -> m (Set a)
deserializeSet sz elem = go sz
  where
    -- go :: Size -> m (Set a)
    go n
      | n <= 0    = pure Tip
      | otherwise = do
           l <- go n_half
           x <- elem
           r <- go (n - n_half - 1)
           pure $! Bin n x l r
        where n_half = n `unsafeShiftR` 1
```

**Comment by treeowl** [comment]
Actually, I think we probably want to shift names and offer two versions:

```haskell
deserialize
  :: Applicative m
  => Size
  -> m a
  -> m (Set a)
deserialize sz elem = go sz
  where
    -- go :: Size -> m (Set a)
    go n
      | n <= 0
      = pure Tip
      | otherwise
      = liftA3 (Bin n)
           (go n_half)
           elem
           (go (n - n_half - 1))
      where n_half = n `unsafeShiftR` 1

deserialize'
  :: Monad m
  => Size
  -> m a
  -> m (Set a)
deserialize' sz elem = go sz
  where
    -- go :: Size -> m (Set a)
    go n
      | n <= 0    = pure Tip
      | otherwise = do
           l <- go n_half
           x <- elem
           r <- go (n - n_half - 1)
           pure $! Bin n x l r
      where n_half = n `unsafeShiftR` 1
```

**Comment by sjakobi** [comment]
Nice! :) I hadn't realized that `Applicative` would be enough!

What's the point of offering an additional `Monad` version though?

**Comment by treeowl** [comment]
Won't the Monad version give better strictness analysis sometimes?

**Comment by treeowl** [comment]
We can certainly start out with just the Applicative version.

**Comment by sjakobi** [comment]
> Won't the Monad version give better strictness analysis sometimes?

I have no idea! :/

An analogous function for `Map`s (`Applicative m => Size -> m (k, v) -> m (Map k v)`) should be easy.

I haven't looked into `IntSet` and `IntMap` yet.

**Comment by sjakobi** [comment]
> Won't the Monad version give better strictness analysis sometimes?

It looks like you were onto something there. In my benchmarks the Monad version is about twice as fast as the Applicative one. The Applicative version is even slightly slower than `fromDistinctAscList <$> deserializeList`.

The difference in the generated core is very small though. Here's the Applicative version specialized to `IO`:

```
$wd
  = \ @ a ww w w1 ->
      letrec {
        $wgo
        $wgo
          = \ ww1 w2 ->
              case <=# ww1 0# of {
                __DEFAULT ->
                  let {
                    n_half
                    n_half = uncheckedIShiftRA# ww1 1# } in
                  case $wgo n_half w2 of { (# ipv, ipv1 #) ->
                  case (w `cast` <Co:2>) ipv of { (# ipv2, ipv3 #) ->
                  case $wgo (-# (-# ww1 n_half) 1#) ipv2 of { (# ipv4, ipv5 #) ->
                  (# ipv4,
                     case ipv3 of dt { __DEFAULT ->
                     case ipv1 of dt1 { __DEFAULT ->
                     case ipv5 of dt2 { __DEFAULT -> Bin ww1 dt dt1 dt2 }
                     }
                     } #)
                  }
                  }
                  };
                1# -> (# w2, Tip #)
              }; } in
      $wgo ww w1
```

The Monad version is nearly the same except the unboxed result tuple in the `Bin` case doesn't wrap the last few case expressions:

```
...
                  case ipv3 of dt { __DEFAULT ->
                  case ipv1 of dt1 { __DEFAULT ->
                  case ipv5 of dt2 { __DEFAULT -> (# ipv4, Bin ww1 dt dt1 dt2 #) }
...
```

I'm not very familiar with GHC's optimizations though. Is this something we should report on GHC's Trac or so?

**Comment by treeowl** [comment]
The compiler can't do any better. The trouble is that it has no way to know that you will actually use the result, and therefore (in particular) it can't force the element values it needs to construct the set eagerly. I'd be more interested in benchmark results from a deserialization monad. I suspect they'll be similar but my intuition isn't always right.

**Comment by treeowl** [comment]
My semi-educated guess is that the way to get the `Applicative` version to optimize reasonably well is something like this:

```haskell
{-# INLINE deserialize #-}
deserialize
  :: Applicative m
  => Size
  -> m a
  -> m (Set a)
deserialize sz elem0 = go sz elem0
  where
    -- go :: Size -> m a -> m (Set a)
    go n elem
      | n <= 0
      = pure Tip
      | otherwise
      = liftA3 (Bin n)
           (go n_half elem)
           elem
           (go (n - n_half - 1) elem)
      where n_half = n `unsafeShiftR` 1
```

Now GHC may (hopefully) be able to use either the fact that the demand is strict or the (possible) fact that the element deserialization produces something in WHNF. It's worth a shot (especially benchmarking with a deserialization `Applicative`), but I think offering a strict version too would be a good idea.
